---
title: "Pathway enrichment analysis"
author: Fiona Dick
output:
  html_document:
    fig_width: 10
    fig_height: 5
    toc: true
    toc_depth: 6
    toc_float: true
    self_contained: true
---



```{r echo = F}
#set JAVA_HOME correctly for ermineR (ermineJ wrapper) to work
Sys.setenv(JAVA_HOME = "/usr/lib/jvm/java-1.11.0-openjdk-amd64/")
Sys.setenv(JAVA_HOME="/usr/lib/jvm/java-11-openjdk-amd64")
unixtools::set.tempdir("./tmp")
library(magrittr)
library(reshape2)
library(ggplot2)
library(viridis)
library(ComplexHeatmap)
library(patchwork)
library(ermineR)
library(igraph)
library(dendextend)
library(colormap)
library(kableExtra)
library("xlsx")
library(colorspace)
library(dplyr)
library(RColorBrewer)
library(parallel)
library(scales)
library(fgsea)
```


```{r echo = F}
knitr::opts_chunk$set(message = F, warning = F, cache.extra = set.seed(1))
gene_cl <- readRDS("./results/rds/gene_cl.rds")
#gene_cl_pd <- readRDS("./results/rds/gene_cl_pd.rds")
out <- "./result_figs/pea/"
outKEGG <- "./result_figs/pea/kegg/"
outGO  <- "./result_figs/pea/go/"
pathways <- list(KEGG = "./referenceData/c2.cp.kegg.v7.5.1.symbols.gmt", GO_simplified = "./referenceData/Data_GO_simplified_0.4.gmt", mito  = "./referenceData/BIOCARTA_MITOCHONDRIA_PATHWAY.v7.5.1.gmt")
source("./functions.R")
```

Seed was set (set.seed(1)) in the knitr options to ensure reproducability for all chunks. Random numer generator should result in:
- 94712531068

```{r}
sample(10)
```

```{r}
cols <- c("#ef476f", "#073b4c", "#118ab2", "#06d6a0", "#84a59d")
names(cols) <- c("PD", "HA", "Other", "YG", "Reference") 
```

```{r}
#gsr_annot <- GetAnnoFiles("Generic_human")
```

# Generate gene scorings for all questions

We want to compare:  

1) YG and HA to identify biological functions and processes that show a change in coupling patterns in normal ageing  
2) YG and PD to identify biological functions and processes that show a change in coupling patterns in ageing with PD
3) Explicitly PD and HA to identify biological functions and processes that show altered coupling patterns in PD

For each comparison we are intrested in the following questions:

a) Which genes decouple compared to the reference
b) Which genes become inversely correlated compared to the reference 
c) Which genes show an increase in positive correlation compared to the reference


For each of these question we create a sorted list of genes and perform functional gene set enrichment analysis and test for enrichement of GO pathways in the high ranking genes. 

For question $i \in {1, 2, 3}$ we calculate $S_{a}^{i}$, $S_{b}^{i}$ and $S_{c}^{i}$:
$$
\begin{equation}
  R_{ref}=\begin{cases}
    R_{YG}, & \text{for i} \in {1, 2}\\
    R_{HA}, & \text{for i = 3}
     \end{cases}
\end{equation} $$

$$
\begin{equation}
  R_{ageing}=\begin{cases}
       R_{PD}, & \text{for i} \in {2,3}\\
	R_{HA}, & \text{for i = 2}
  \end{cases}
\end{equation} $$
$$\begin{align}
\forall R_{ref} > 0\\
	S_{a}^{i} &= -|R_{ageing}| + R_{ref}\\
	S_{b}^{i} &= -R_{ageing} + t(R_{ref})\\ 
	S_{c}^{i} &= R_{ageing} - t(R_{ref}),\\
	\textit{with} \ t(x) = \frac{x + 1}{2}
\end{align}$$

To investigate differences in results between two comparisons $i$ and $j$, we rank genes according to the delta $\delta$ in each of the three questions:  
$$\begin{align}
\forall R_{YG} > 0\\
\delta S_{a} &= -S^{i}_{a} +  S^{j}_{a} \\
\delta S_{b} &= -S^{i}_{b} +  S^{j}_{b} \\
\delta S_{c} &= -S^{i}_{c} +  S^{j}_{c}
\end{align}
$$


```{r, echo = F}
#moves correlation to positive (like rescale)
trans <- function(x)
{	
	(x+1)/2
}
```

```{r echo = F}
score_map <- function(df, score_name, ref, ageing, outdir = out){
	df %<>% 
		dplyr::filter(!is.na(!!as.symbol(score_name)))  
	rownames(df)  <- df$gene_name
	col_fun1  <-  circlize::colorRamp2(c(-1,0,1), viridis(3))
	col_fun2  <- circlize::colorRamp2(range(df[, score_name]), c("white", "#4287f5"))
	ha  <-  HeatmapAnnotation("Ageing"  = as.numeric(unlist(df[, ageing])),
			  "Reference" = as.numeric(unlist(df[, ref])),
			col = list("Ageing" = col_fun1,
				   "Reference"  = col_fun1),
			show_legend = c(T, F),
			annotation_name_side = "left",
			annotation_legend_param = list("Ageing" = list(title = "R:\n rna ~ prot")))
	p <- ComplexHeatmap::Heatmap(t(df %>% 
				       dplyr::ungroup() %>%
				dplyr::select(score_name)),
			top_annotation =  ha,
			height = unit(.7, "cm"),
			show_column_names = F ,
			show_row_names = F,
			name = paste0("Ranking:\n", score_name),
			col = col_fun2,
			column_dend_height = unit(20, "mm"))
	pdf(paste0(out, score_name, ".pdf"))
	draw(p)
	dev.off()
	p <- ggplotify::as.ggplot(grid.grabExpr(draw(p)))
	return(p)
}
```

## Comparison 1 : HA and YG

- Reference group = YG
- Ageing group = HA

Of the positively correlated genes in YG:
	- a) which decouple with HA
	- b) which anti-correlate with HA?
	- c) which increase in correlation in HA?

 Reducing the number of genes tested here, since we filter for genes that 
 show positive correlation in YG (our reference).


```{r}
gene_cl %>% dplyr::filter(YG > 0) %>%
	dplyr::mutate(S1_a = -abs(HA) + YG,
		      S1_b = -HA + trans(YG),
		      S1_c = HA - trans(YG)) -> df_scorings
```

Plot the gene score distribution

```{r echo = F}
p_df <- melt(df_scorings %>%
	     dplyr::filter(YG > 0) %>%
	     dplyr::select(gene_name, HA, YG, S1_a, S1_b, S1_c), id.vars = c("gene_name", "HA", "YG")) %>%
	dplyr::rename(Question = variable, Score = value) %>%
	melt(., id.vars = c("gene_name", "Question"))
p1 <- ggplot(p_df, aes(x = value, col = variable)) +
	geom_density(aes(y = ..density..)) +
	scale_color_manual(values = c("HA" = cols[["HA"]], "YG" = cols[["YG"]], "Score" = "#4287f5")) +
	theme_light() +
	geom_rug(data = p_df %>% dplyr::filter(variable %in% c("Score")), aes(x = value, col = variable), sides = "b") +
	facet_wrap(~Question, scales = "free", ncol = 1) +
	theme(legend.position = "bottom") +
	labs(x = "Value", y = "Density", col = "Variable") 
p1a <- score_map(df_scorings,
		score_name = "S1_a",
		ref = "YG",
		ageing = "HA",
		outdir = out)
p1b <- score_map(df_scorings,
		score_name = "S1_b",
		ref = "YG",
		ageing = "HA",
		outdir = out)
p1c <- score_map(df_scorings,
		score_name = "S1_c",
		ref = "YG",
		ageing = "HA",
		outdir = out)

the_plot_1 <- p1 | (p1a / p1b / p1c)
```

## Comparison 2 : PD and YG

- Reference group = YG
- Ageing group = PD

Of the positively correlated genes in YG:
	- a) which decouple with PD
	- b) which anti-correlate with PD?
	- c) which increase in correlation in PD?

```{r}
df_scorings %<>%
		dplyr::mutate(S2_a = -abs(PD) + YG,
			      S2_b = -PD + trans(YG),
			      S2_c = PD - trans(YG)) 
```


```{r}

p_df <- melt(df_scorings %>%
	     dplyr::filter(YG > 0, !is.na(PD)) %>%
	     dplyr::select(gene_name, PD, YG, S2_a, S2_b, S2_c), id.vars = c("gene_name", "PD", "YG")) %>%
	dplyr::rename(Question = variable, Score = value) %>%
	melt(., id.vars = c("gene_name", "Question"))
p1 <- ggplot(p_df, aes(x = value, col = variable)) +
	geom_density(aes(y = ..density..)) +
	scale_color_manual(values = c("PD" = cols[["PD"]], "YG" = cols[["YG"]], "Score" = "#4287f5")) +
	theme_light() +
	geom_rug(data = p_df %>% dplyr::filter(variable %in% c("Score")), aes(x = value, col = variable), sides = "b") +
	facet_wrap(~Question, scales = "free", ncol = 1) +
	labs(x = "Value", y = "Density", col = "Variable")
p1a <- score_map(df_scorings,
		score_name = "S2_a",
		ref = "YG",
		ageing = "PD",
		outdir = out)
p1b <- score_map(df_scorings,
		score_name = "S2_b",
		ref = "YG",
		ageing = "PD",
		outdir = out)
p1c <- score_map(df_scorings,
		score_name = "S2_c",
		ref = "YG",
		ageing = "PD",
		outdir = out)

the_plot_2 <- p1 | (p1a / p1b / p1c)
```

## Comparison 3 : PD and HA

- Reference group = HA
- Ageing group = PD

Of the positively correlated genes in HA:
	- a) which decouple with PD
	- b) which anti-correlate with PD?
	- c) which increase in correlation in PD?

```{r}
df_scorings %<>%
	dplyr::mutate(S3_a = ifelse(HA > 0, -abs(PD) + HA, NA),
		      S3_b = ifelse(HA > 0,-PD + trans(HA), NA),
		      S3_c = ifelse(HA > 0, PD - trans(HA), NA)) 
```

Plot the gene score distribution

```{r echo = F}
p_df <- melt(df_scorings %>%
	     dplyr::filter(HA > 0) %>%
	     dplyr::select(gene_name, PD, HA, S3_a, S3_b, S3_c), id.vars = c("gene_name", "PD", "HA")) %>%
	dplyr::rename(Question = variable, Score = value) %>%
	melt(., id.vars = c("gene_name", "Question"))
p1 <- ggplot(p_df, aes(x = value, col = variable)) +
	geom_density(aes(y = ..density..)) +
	scale_color_manual(values = c("PD" = cols[["PD"]], "HA" = cols[["HA"]], "Score" = "#4287f5")) +
	theme_light() +
	geom_rug(data = p_df %>% dplyr::filter(variable %in% c("Score")), aes(x = value, col = variable), sides = "b") +
	facet_wrap(~Question, scales = "free", ncol = 1) +
	labs(x = "Value", y = "Density", col = "Variable")
p1a <- score_map(df_scorings,
		score_name = "S3_a",
		ref = "HA",
		ageing = "PD",
		outdir = out)
p1b <- score_map(df_scorings,
		score_name = "S3_b",
		ref = "HA",
		ageing = "PD",
		outdir = out)
p1c <- score_map(df_scorings,
		score_name = "S3_c",
		ref = "HA",
		ageing = "PD",
		outdir = out)
the_plot_3 <- p1 | (p1a / p1b / p1c)
```

## Pathway enrichment analysis for each gene score


### wanted to reproduce results for generating suppl. material (18/03/21)
ermineR doesnt work anymore
cannot download ontology from http://archive.geneontology.org/latest-termdb/go_daily-termdb.rdf-xml.gz
Fromat rdf is deprecated.
In this issue I found a "recent" file that I could use.
https://github.com/geneontology/helpdesk/issues/302
This was uploaded by the ermineR developer here: https://www.dropbox.com/s/05i1qrc30dedoqr/go_daily-termdb.rdf-xml.gz?dl=0
It is now in ./referenceData, was unziped and and will be used for the arg geneSetDescription until they fix ermineR.

Edit:
ermineR works with .rdf-xml file as described above in version ermineR_1.0.1.9000, with gemmaAPI_2.0.3.9000, but when I ran this on azure it didnt work, then i used the latest .obo file, same ermineR version but gemmaAPI_2.0.4.9000


Edit:
Changed to fgsea instead, easier to use and install

```{r}
do_gsr <-function(df, score_name, which_alpha = 0.05, tool = "fgsea",
 bigIsBetter = TRUE, iterations = 200000, tdir = out,
 return_id = F, custom_geneset = NULL, append_excel = TRUE,
 cluster = F, perms = F, repl = "", pretty = F, selection = NULL) {

 df <- as.data.frame(df)
 rownames(df)  <- df$gene_name
 df <- df %>% dplyr::select(gene_name, all_of(score_name)) %>%
	na.omit(.)
#Run enrichment with either fgsea or ermineR, if fgsea, need also geneset annotation file
 if(tool == "fgsea" & !is.null(custom_geneset)) {
	custom_geneset <- fgsea::gmtPathways(custom_geneset)
	if(!is.null(selection)) {
		custom_geneset <- custom_geneset[names(custom_geneset) %in% selection]
	}
	df %<>% dplyr::ungroup(.)
	ranks <- setNames(df %>% dplyr::pull(all_of(score_name)),
			  df %>% dplyr::pull(gene_name))
	#Added this after switching to fgsea to shift scores to positive
 	ranks <- rescale(ranks)
	ranks <- sort(ranks, decreasing = T)
	res_all   <- fgsea::fgseaMultilevel(pathways = custom_geneset,
			        stats = ranks,
				sampleSize = 10000,
				nproc = 9,
				scoreType = "pos",
				minSize = 10,
				maxSize = 200) 
	res_all %>% dplyr::arrange(padj) %>%
		dplyr::filter(padj < which_alpha) -> res 
} else {
 ermineR::gsr(scores = df,
	        geneReplicates = "best",
		pAdjust = "FDR",
		iterations = iterations,
		minClassSize = 5,
		maxClassSize = 250,
         	scoreColumn = score_name,
		bigIsBetter = bigIsBetter,
	     	annotation = gsr_annot,
		geneSetDescription  = "./referenceData/go_daily-termdb.rdf-xml",
	     	aspects = c("C", "M", "B"))$results -> pea
 
 res <- pea %>% dplyr::arrange(CorrectedPvalue) %>%
	dplyr::filter(CorrectedPvalue < which_alpha)%>%
	dplyr::select(ID, CorrectedPvalue, Name, GeneMembers, NumGenes)
} 
#If perms == T, meaning if function is called withing apply that goes through permutations, r
# rename the score collumn so that resulting list can be merged
if(perms == T & tool == "fgsea") {
	#print(paste0("IM here: ", repl))
	res_all %>% 
	dplyr::select(pathway, ES) %>%
	data.table::as.data.table(.) -> res_all
 	data.table::setkeyv(res_all, c("pathway")) 
	data.table::setnames(res_all,
			     old = "ES",
			     new = paste0("ES_", repl)) 
	return(res_all)
}
  DT::datatable(res,
		extensions = 'Buttons',
		options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print')))
  if(pretty) {
	res_all %>% 
		dplyr::filter(padj < which_alpha) %>%
		dplyr::arrange(padj) %>%
		dplyr::mutate(groupsCompared = case_when(grepl("S1", score_name) ~ "YG -> HA",
							 grepl("S2", score_name) ~ "YG -> PD",
							 grepl("S3", score_name) ~ "HA -> PD")) %>%
		dplyr::mutate(scoreType = case_when(grepl("a", score_name) ~ "Decoupling",
						    grepl("b", score_name) ~ "Incr. neg. correlation",
						    grepl("c", score_name) ~ "Incr. pos. correlation")) %>%
		dplyr::mutate(score_name = rep(score_name, nrow(.))) %>%
		dplyr::select(score_name, groupsCompared, scoreType, pathway, padj, ES) -> res
	return(res)
  }
  return(res_all %>% dplyr::arrange(padj)) 
 
}
```

### Question 1: Changes in coupling in healthy ageing

```{r echo = F}
the_plot_1
```


#### S1 Decoupling:

```{r}
S1_a_kegg <- do_gsr(df_scorings, "S1_a",
       which_alpha = 0.05,
       append_excel = F,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S1_a_go  <- do_gsr(df_scorings, "S1_a",
       which_alpha = 0.05,
       append_excel = F,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
```

We found no enrichment for genes the decouple with HA.
We were interested whether the genes are connected in a protein interaction network based on evidence from STRING from:
 - experimental experimental evidence
 - neighborhood
 - coexpression 
 

To investigate the decoupled genes in this manner, we look at the score distribution and choose a hard threshold,
to select the top genes of S1a ranking. 

```{r echo = F}
quantile(df_scorings$S1_a)
cutoff = quantile(df_scorings$S1_a, probs = 0.9)
```

We choose the cutoff = 0.5, which is approx.
the third quantile of S1_a, to select all genes for which S1_a > cutoff for the PPI of decoupled genes

--> edit increasedthe cutoff to 0.9, to have fewer nodes in the network

```{r echo = F} 
library(igraph)
library(coexnet)


S1_a_ppi_df <- df_scorings %>%
	dplyr::filter(S1_a > cutoff) 
ID <- S1_a_ppi_df %>% 
	dplyr::pull(gene_name)
ppi <- ppiNet(molecularIDs = ID,evidence = c("neighborhood","coexpression","experiments"))
S1_a_ppi_df <- S1_a_ppi_df %>%
	dplyr::filter(gene_name %in% names(V(ppi)))
col_fun1 <- colorRampPalette(c("white", "#4287f5"))
graphCol = col_fun1(300)[as.numeric(cut(S1_a_ppi_df$S1_a,breaks = 200))] 
```
Little n analysis

```{r echo = F} 
edge_density(ppi, loops=F)
# transitivity 
transitivity(ppi, type="global")
# (length of the shortest path between two nodes)
diameter(ppi, directed=F, weights=NA)
deg <- degree(ppi, mode="all")
hist(deg, breaks=1:vcount(ppi)-1, main="Histogram of node degree")
# For undirected matrices the adjacency matrix is symmetric and the hub scores are the same as authority scores, see authority_score.
hs <- hub_score(ppi, weights=NA)$vector
hs_color  <- rep("black", length(hs)) 
hs_color[which(hs > 0.8)] <- "violetred2"
# Community detection based on edge betweenness (Newman-Girvan)
# High-betweenness edges are removed sequentially (recalculating at each step) and the best partitioning of the network is selected.
ceb <- cluster_edge_betweenness(ppi) 
```


```{r echo = F}
library(fields)
set.seed(1)
deg <- degree(ppi, mode="all")
ppi <- set_vertex_attr(ppi, "Score", index = V(ppi), value = S1_a_ppi_df$S1_a)
legend_info <- data.frame("Score" = S1_a_ppi_df$S1_a,
			  "col" = graphCol) %>%
	dplyr::arrange(desc(Score))
legend_image <- as.raster(matrix(legend_info$col, ncol=1))
wc <- cluster_edge_betweenness(ppi)
mark <- membership(wc)
mark <- communities(wc)[which(sapply(communities(wc), length) > 5)]
set.seed(140)
pdf("./result_figs/Draft/ppi_ha.pdf", width = 10, height = 10)
p <- plot(ppi, 
     frame = T,
     mark.groups = mark,
     mark.shape = 0.2,
     mark.col = rep("lightgrey", length(mark)),  
     mark.border = "black",
     vertex.color = graphCol,
     vertex.label.cex = .8,
     vertex.frame.width = 10,
     edge.color = "black",
#     edge.curved= seq(-0.5, 0.5, length = ecount(ppi)),
#     loop.angle = 120,
     vertex.label.dist=.8,
     vertex.size = 1.5*deg,
     layout = layout_nicely)
rasterImage(legend_image, -1.1, -1.2, -1, -.5 )
text(x = - 0.95,
     y = c(-0.51, -0.85 , -1.2),	   
     labels = c(paste0("-", round(max(legend_info$Score), digit = 2)),
		paste0("-", round(median(legend_info$Score), digit = 2)),
		paste0("-", round(min(legend_info$Score), digit = 2))))
dev.off()
```



#### S1 Anti-correlation:  

```{r}
S1_b_kegg <- do_gsr(df_scorings, "S1_b",
       which_alpha = 0.05,
       append_excel = F,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S1_b_go <- do_gsr(df_scorings, "S1_b",
       which_alpha = 0.05,
       append_excel = F,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
S1_b_go
S1_b_kegg
```

#### S1 Increase in correlation:  

```{r}
S1_c_kegg <- do_gsr(df_scorings, "S1_c",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S1_c_go <- do_gsr(df_scorings, "S1_c",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
S1_c_go
S1_c_kegg
```


### Question 2: Changes in coupling in ageing with PD  

```{r echo = F}
the_plot_2
```


#### S2 Decoupling:  

```{r}
S2_a_kegg <- do_gsr(df_scorings, "S2_a",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S2_a_go <- do_gsr(df_scorings, "S2_a",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
```

#### PPI on decoupled in PD


```{r echo = F}
quantile(df_scorings$S2_a)
cutoff = quantile(df_scorings$S2_a, probs = 0.9)
```

We choose the cutoff = 0.5, which is approx.
the third quantile of S1_a, to select all genes for which S1_a > cutoff for the PPI of decoupled genes

--> edit increasedthe cutoff to 0.9, to have fewer nodes in the network

```{r echo = F} 
library(igraph)
library(coexnet)


S2_a_ppi_df <- df_scorings %>%
	dplyr::filter(S2_a > cutoff) %>%
	dplyr::arrange(S2_a)
ID <- S2_a_ppi_df %>% 
	dplyr::pull(gene_name)
ppi <- ppiNet(molecularIDs = ID, evidence = c("neighborhood","coexpression","experiments"))
# filter df to only have genes for which "evidence" in string was found
S2_a_ppi_df <- S2_a_ppi_df %>%
	dplyr::filter(gene_name %in% as.character(names(V(ppi))))
col_fun1 <- colorRampPalette(c("white", "#4287f5"))
graphCol = col_fun1(300)[as.numeric(cut(S2_a_ppi_df$S2_a,breaks = 200))] 
```
Little n analysis

```{r echo = F} 
edge_density(ppi, loops=F)
# transitivity 
transitivity(ppi, type="global")
# (length of the shortest path between two nodes)
diameter(ppi, directed=F, weights=NA)
deg <- degree(ppi, mode="all")
hist(deg, breaks=1:vcount(ppi)-1, main="Histogram of node degree")
# For undirected matrices the adjacency matrix is symmetric and the hub scores are the same as authority scores, see authority_score.
hs <- hub_score(ppi, weights=NA)$vector
hs_color  <- rep("black", length(hs)) 
hs_color[which(hs > 0.8)] <- "violetred2"
# Community detection based on edge betweenness (Newman-Girvan)
# High-betweenness edges are removed sequentially (recalculating at each step) and the best partitioning of the network is selected.
ceb <- cluster_edge_betweenness(ppi) 
```


```{r echo = F}
library(fields)
set.seed(1)
deg <- degree(ppi, mode="all")
ppi <- set_vertex_attr(ppi, "Score", index = V(ppi), value = S2_a_ppi_df$S2_a)
legend_info <- data.frame("Score" = S2_a_ppi_df$S2_a,
			  "col" = graphCol) %>%
	dplyr::arrange(desc(Score))
legend_image <- as.raster(matrix(legend_info$col, ncol=1))
wc <- cluster_edge_betweenness(ppi)
mark <- membership(wc)
mark <- communities(wc)[which(sapply(communities(wc), length) > 5)]
set.seed(140)
pdf("./result_figs/Draft/ppi_pd.pdf", width = 10, height = 10)
p <- plot(ppi, 
     frame = T,
     mark.groups = mark,
     mark.shape = 0.2,
     mark.col = rep("lightgrey", length(mark)),  
     mark.border = "black",
     vertex.color = graphCol,
     vertex.label.cex = .8,
     vertex.frame.width = 10,
     edge.color = "black",
#     edge.curved= seq(-0.5, 0.5, length = ecount(ppi)),
#     loop.angle = 120,
     vertex.label.dist=.8,
     vertex.size = 1.5*deg,
     layout = layout_nicely)
rasterImage(legend_image, -1.1, -1.2, -1, -.5 )
text(x = - 0.95,
     y = c(-0.51, -0.85 , -1.2),	   
     labels = c(paste0("-", round(max(legend_info$Score), digit = 2)),
		paste0("-", round(median(legend_info$Score), digit = 2)),
		paste0("-", round(min(legend_info$Score), digit = 2))))
dev.off()
```


#### S2 Anti-correlation:  

```{r}
S2_b_kegg <- do_gsr(df_scorings, "S2_b",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S2_b_go <- do_gsr(df_scorings, "S2_b",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
S2_b_go
S2_b_kegg
```

#### S2 Increase in correlation:  

```{r}
S2_c_kegg <- do_gsr(df_scorings, "S2_c",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S2_c_go <- do_gsr(df_scorings, "S2_c",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
```

### Question 3: Changes in coupling from HA to PD  

```{r echo = F}
the_plot_3
```


#### S3 Decoupling:  

```{r}
S3_a_kegg <- do_gsr(df_scorings, "S3_a",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S3_a_go <- do_gsr(df_scorings, "S3_a",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
S3_a_kegg
```

#### S3 Anti-correlation:  

```{r}
S3_b_kegg <- do_gsr(df_scorings, "S3_b",
       which_alpha = 0.05,
       append_excel = F,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S3_b_go <- do_gsr(df_scorings, "S3_b",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
S3_b_kegg
```

#### S3 Increase in correlation:   

```{r}
S3_c_kegg <- do_gsr(df_scorings, "S3_c",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["KEGG"]],
       tdir = outKEGG,
	pretty = T)
S3_c_go <- do_gsr(df_scorings, "S3_c",
       which_alpha = 0.05,
       append_excel = T,
       tool = "fgsea",
       custom_geneset = pathways[["GO_simplified"]],
       tdir = outGO,
	pretty = T)
```


### Summarize results in table
(Table 1)

```{r}
df_pathways <- do.call(rbind, list(S1_a_go, S1_a_kegg,
		    S1_b_go, S1_b_kegg,
		    S1_c_go, S1_c_kegg,
		    S2_a_go, S2_a_kegg,
		    S2_b_go, S2_b_kegg,
		    S2_c_go, S2_c_kegg,
		    S3_a_go, S3_a_kegg,
		    S3_b_go, S3_b_kegg,
		    S3_c_go, S3_c_kegg)) 
df_pathways
```


#### Permuted dataset

```{r}
run_gsea_perms <- function(df_scorings_perm, score_col, which_ont, sig_pathways = NULL) {
	cl = parallel::makeCluster(5, type = "FORK")
	pbapply::pblapply(replicates, function(repl_i) {				  
		df_scorings_perm %>% dplyr::filter(replicate == repl_i) %>%
		dplyr::ungroup(.) %>%
		dplyr::select(gene_name, all_of(score_col)) %>%
		na.omit(.) %>%
		as.data.frame(.) -> df
		do_gsr(df = df,
		       score_name = score_col,
		       custom_geneset = pathways[[which_ont]],
		       perms = T,
		       repl = repl_i,
		       tool = "fgsea",
		       selection = sig_pathways)
	}) -> df
		
	reduce(df, full_join, by = "pathway") -> df
		print(head(df))
	stopCluster(cl)
	df %>%
		dplyr::rename(obs = paste0("ES_", max(replicates))) %>% 
		dplyr::select(pathway, obs, contains("ES_")) -> df  
	tdate = Sys.Date()
	saveRDS(df, paste0("./rdsData/perms/", score_col,"_perm_5000_", tdate, "_" , which_ont,".rds"))
	#print(head(df))
	return(df)
}
```
--> The following code is commented out to not accidentally run it (takes a long time)

#### PEA on permuations

To run pathway enrichment analysis on the permuted dataset (created in analysis.Rmd) 
We read in the permuted dataset and calculate gene scores as in our analysis above.
Reading in the rds file again instead of saving it to a obj to save RAM. Probably smarter ways to do this

```{r}
nperm = 5000
readRDS("./rdsData/perms/gene_cl_perm_9_15_4_5000.rds") %>%
	dplyr::rename(gene_name = gene_name_u) %>% 
	dplyr::filter(k_4 > 0) %>% 
	dplyr::mutate(S1_a = -abs(k_9) + k_4,
		      S1_b = -k_9 + trans(k_4),
		      S1_c = k_9 - trans(k_4)) %>%
	dplyr::bind_rows(., (df_scorings %>%
			 dplyr::select(gene_name, S1_a, S1_b, S1_c) %>%
			 dplyr::mutate(replicate = (nperm + 1)))) -> df_scorings_perm_S1

readRDS("./rdsData/perms/gene_cl_perm_9_15_4_5000.rds") %>%
	dplyr::rename(gene_name = gene_name_u) %>%
	dplyr::filter(k_4 > 0) %>% 
	dplyr::mutate(S2_a = -abs(k_15) + k_4,
		      S2_b = -k_15 + trans(k_4),
		      S2_c = k_15 - trans(k_4)) %>%
	dplyr::bind_rows(., (df_scorings %>%
			 dplyr::select(gene_name, S2_a, S2_b, S2_c) %>%
			 dplyr::mutate(replicate = (nperm + 1)))) -> df_scorings_perm_S2


readRDS("./rdsData/perms/gene_cl_perm_9_15_4_5000.rds") %>%
	dplyr::rename(gene_name = gene_name_u) %>%
	dplyr::filter(k_9 > 0) %>% 
	dplyr::mutate(S3_a = -abs(k_15) + k_9,
		      S3_b = -k_15 + trans(k_9),
		      S3_c = k_15 - trans(k_9)) %>% 
	dplyr::bind_rows(., (df_scorings %>%
			 dplyr::select(gene_name, S3_a, S3_b, S3_c) %>%
			 dplyr::mutate(replicate = (nperm + 1)))) -> df_scorings_perm_S3



replicates = seq(1, (nperm + 1),1)
print(replicates)
```

#### Running fgsea on permutations only for the score types that showed sig. enrichment

To reduce runtime Im running the pea only on the pathways that we want to do permutation test for
Synaptic vesicle in S1_b

```{r}
library(parallel)
library(tidyverse)

selection  <- df_pathways %>%
		dplyr::filter(score_name == "S1_b") %>%
		dplyr::pull(pathway)
run_gsea_perms(df_scorings_perm_S1, "S1_b", "GO_simplified", sig_pathways = selection) 
```

Check proteasome in S2_b KEGG and the other sig. pathway in GO

```{r}
selection  <- df_pathways %>%
		dplyr::filter(score_name == "S2_b") %>%
		dplyr::pull(pathway) 

S2_b_perm_KEGG <- run_gsea_perms(df_scorings_perm_S2, "S2_b", "KEGG", sig_pathways = selection)
S2_b_perm_GO <- run_gsea_perms(df_scorings_perm_S2, "S2_b", "GO_simplified", sig_pathways = selection)
```

Check proteasome in S3_b KEGG and the other sig. pathway in GO

```{r}
#S3_b_perm_GO <- run_gsea_perms(df_scorings_perm_S3, "S3_b", "GO_simplified")
selection  <- df_pathways %>%
		dplyr::filter(score_name == "S3_b") %>%
		dplyr::pull(pathway) 
#upnext
run_gsea_perms(df_scorings_perm_S3, "S3_b", "KEGG", sig_pathways = selection)
```

Function to evaluate the permuation results
If you rerun this, remember to changed the name of the result file name you read in
by adding the date you reran the pea analysis. Added date to not overwrite files used for the manuscript

```{r}
perm_res <- function(score_col, which_ont) {

filep = paste0("./rdsData/perms/", score_col, "_perm_5000_", which_ont, ".rds")
df = readRDS(filep)
head(df)
nperm = 5000

df %>% 	#IMPORTANT: we calculated gene specific correlations after sampling samples in each permutation
	# Obv. this results in different correlation coefficients for the same gene and group in different permutations
	# Pea so the increased anti correlation score is done by filtering genes that are above zero in the reference group (Figure 1)
	# So in the permutations different set of genes will be filtered
	# This can lead to some pathways not appearing in the fgsea result in some permutations
	# Here we set the ES of these pathways to zero, as they are not enriched
	# It is important to note that for each comparison and permutation we apply exactly the same rules for the analysis as for the "real" values
	dplyr::mutate_at(vars(contains('ES_')),funs(ifelse(is.na(.), 0, .))) %>%
	dplyr::rowwise() %>%
  	dplyr::mutate(p_res = ifelse(is.na(obs), NA, 
#TODO include ifelse to fifferentiate between left or right side of median
				       ifelse(obs > median(c_across(contains("ES_") ), na.rm = T),
				       		sum(c_across(contains("ES_")) > obs, na.rm = T) / (dim(df)[2]-2
),
						sum(c_across(contains("ES_")) < obs, na.rm = T) / (dim(df)[2]-2
)))) %>%
	dplyr::mutate(quantile95 = quantile(c_across(contains("ES_")), probs = .95, na.rm = T)) %>%
	dplyr::ungroup() %>%
	#	dplyr::mutate(padj = p.adjust(p_res)) %>%
#	dplyr::filter(p_res<0.05) %>%
	dplyr::arrange(p_res) %>%
	dplyr::mutate(score_name = rep(score_col, nrow(.)),
		      Ontology = rep(which_ont, nrow(.))) %>%
	dplyr::select(pathway, p_res, score_name, Ontology) 
}

#Calculate permutation p value for the results we were interessted in:
perm_res_S1_b <- perm_res("S1_b", "GO_simplified")
perm_res_S2_b_KEGG  <- perm_res("S2_b", "KEGG")
perm_res_S2_b_GO  <-  perm_res("S2_b", "GO_simplified")
perm_res_S3_b  <- perm_res("S3_b", "KEGG")
# Bind results to a df
df <- do.call(rbind, list(perm_res_S1_b,
			  perm_res_S2_b_GO,
			  perm_res_S2_b_KEGG,
			  perm_res_S3_b))
dplyr::right_join(df, df_pathways, by = c("pathway", "score_name")) %>%
	dplyr::select(groupsCompared, score_name, scoreType, pathway, padj, ES, p_res) %>%
	dplyr::rename(perm_p = p_res) -> tab
write.table(tab, "./result_figs/pea/Table1.txt", sep = ",", quote = F, row.names = F)
xtable::xtable(tab)

```

## Pea analysis on correlation coefficients for each group

i.e. for each group check enrichment for genes that are most negativley correlated and most positivley correlated
Without comparing between groups (Table S3)

```{r}
df <- gene_cl %>% 
	dplyr::mutate(PD_neg = -PD,
		      HA_neg = -HA,
		      YG_neg = -YG)
pea_results <- lapply(colnames(df)[-1], function(score) {
kegg = do_gsr(df, score, which_alpha = 0.05, tool = "fgsea", 
       custom_geneset = pathways[["KEGG"]],
       tdir = outGO,
       pretty = T) %>%
	dplyr::select(-scoreType, -groupsCompared) %>% 
 dplyr::mutate(Ontology = rep("KEGG", nrow(.)))
go = do_gsr(df, score, which_alpha = 0.05, tool = "fgsea",
     custom_geneset = pathways[["GO_simplified"]],
     tdir = outGO,
     pretty = T) %>%
 dplyr::select(-scoreType, -groupsCompared) %>%
 dplyr::mutate(Ontology = rep("GO_simplified", nrow(.)))
return(rbind(kegg, go))
	})
names(pea_results) <- colnames(df)[-1]
df_pea_r <- do.call(rbind, pea_results) 
df_pea_r
```

### Permuations

Let's confirm those results by doing the same enrichment analysis 
on correlation coefficicents that were calculated from randomly sampled subjects
where we made groups of the same size as our originial groups (Nyg = 4, Nha = 9, Npd = 15)

Using the same permutation dataset as for the previous permutation analysis

```{r}
nperm = 5000
readRDS("./rdsData/perms/gene_cl_perm_9_15_4_5000.rds") %>% 
	dplyr::mutate(k15_neg = -k_15, k9_neg = -k_9, k4_neg = -k_4,
		      k15 = k_15, k4 = k_4, k9 = k_9) %>%
	dplyr::mutate(gene_name = gene_name_u) %>%
	dplyr::select(gene_name, replicate, k4, k9, k15, k15_neg, k9_neg, k4_neg) %>%
	dplyr::bind_rows(., (gene_cl %>%
			     dplyr::ungroup(.) %>%
			     dplyr::mutate(k4_neg = -YG,
					   k9_neg = -HA,
					   k15_neg = -PD,
					   k4 = YG,
					   k9 = HA,
					   k15 = PD) %>%
			     dplyr::select(gene_name, k4, k9, k15, k4_neg, k9_neg, k15_neg) %>%
			     dplyr::mutate(replicate = nperm + 1))) -> df

replicates = seq(1,nperm+1,1)
head(df)
# commented the run_gsea_perms lines, calculations are done and saved to rds
# If there is an error when rerunning this, it could be because I had merged two files into one
# and there might be small changes in the run_gsea_perms function but just related to what is returned or columnames
# should be fixable easily 

# PD k15 pos and neg:

#k15negKegg  <- run_gsea_perms(df, "k15_neg", "KEGG")
k15negKegg  <- readRDS("./rdsData/perms/directr/k15_neg_perm_5000KEGG.rds")

#k15Kegg <- run_gsea_perms(df, "k15", "KEGG")
k15Kegg <- readRDS("./rdsData/perms/directr/k15_perm_5000KEGG.rds")

#k15negGo  <- run_gsea_perms(df, "k15_neg", "GO_simplified")
k15negGo  <- readRDS("./rdsData/perms/directr/k15_neg_perm_5000GO_simplified.rds")

#k15Go <- run_gsea_perms(df, "k15", "GO_simplified")
k15Go <- readRDS("./rdsData/perms/directr/k15_perm_5000GO_simplified.rds")


# HA k9 only neg:

#k9negKegg  <- run_gsea_perms(df, "k9_neg", "KEGG")
k9negKegg <- readRDS("./rdsData/perms/directr/k9_neg_perm_5000KEGG.rds")

#k9negGo  <- run_gsea_perms(df, "k9_neg", "GO_simplified")
k9negGo <- readRDS("./rdsData/perms/directr/k9_neg_perm_5000GO_simplified.rds")
```

Calculate pvale and arrange to a table

```{r}
perm_res <- function(score_col, which_ont, obj) { 
	obj %>% 
	dplyr::mutate_at(vars(contains('ES_')),funs(ifelse(is.na(.), 0, .))) %>%
	dplyr::rowwise() %>%
  	dplyr::mutate(p_res = ifelse(is.na(obs), NA, 
#TODO include ifelse to differentiate between left or right side of median
				       ifelse(obs > median(c_across(contains("ES_") ), na.rm = T),
				       		sum(c_across(contains("ES_")) > obs, na.rm = T) / (dim(obj)[2]-2
),
						sum(c_across(contains("ES_")) < obs, na.rm = T) / (dim(obj)[2]-2
)))) %>%
	dplyr::mutate(quantile95 = quantile(c_across(contains("ES_")), probs = .95, na.rm = T)) %>%
	dplyr::ungroup() %>%
	dplyr::arrange(p_res) %>%
	dplyr::mutate(score_name = rep(score_col, nrow(.)),
		      Ontology = rep(which_ont, nrow(.))) %>%
	dplyr::select(pathway, p_res, score_name, Ontology) 
}

# Calculate pvalue and join all the scores together to a df

df_pea_r %>%
	#-HA KEGG and permuations pval from group k9 (9 random samples, genes sorted by -r)
	dplyr::left_join(perm_res("k9_neg", "KEGG", k9negKegg) %>% dplyr::mutate(score_name = "HA_neg"),
			 by = c("score_name", "pathway", "Ontology")) %>%
	#-HA GO (k9_negGO)
	dplyr::left_join(perm_res("k9_neg", "GO_simplified", k9negGo) %>% dplyr::mutate(score_name = "HA_neg"),
			 by = c("score_name", "pathway", "Ontology")) %>%

	#-PD KEGG and permuations pval from group k15 (15 random samples, genes sorted by -r)
	dplyr::left_join(perm_res("k15_neg", "KEGG", k15negKegg) %>% dplyr::mutate(score_name = "PD_neg"),
			 by = c("score_name", "pathway", "Ontology")) %>%
	#-PD GO (k15_negGO)
	dplyr::left_join(perm_res("k15_neg", "GO_simplified", k15negGo) %>% dplyr::mutate(score_name = "PD_neg"),
			 by = c("score_name", "pathway", "Ontology")) %>%

	#+PD KEGG and permuations pval from group k15 (15 random samples, genes sorted by r)
	dplyr::left_join(perm_res("k15", "KEGG", k15Kegg) %>% dplyr::mutate(score_name = "PD"),
			 by = c("score_name", "pathway", "Ontology")) %>%
	#+PD GO (k15GO)
	dplyr::left_join(perm_res("k15", "GO_simplified", k15Go) %>% dplyr::mutate(score_name = "PD"),
			 by = c("score_name", "pathway", "Ontology")) -> tab
foo <- function(x) {
return(x[!is.na(x)])
}
tab %>% 
	dplyr::rowwise() %>%
	dplyr::mutate(p.res = foo(across(6:11))) %>%
	dplyr::ungroup() %>%
	dplyr::select(!starts_with("p_res.")) %>%
write.table(., "./result_figs/pea/Table2.txt",
	    quote = F,
	    row.names = F,
	    sep = ",")
```




### Transcript - protein corrlation of Mitochondrial genes in PD

```{r}
mrc <- readr::read_tsv("./referenceData/MRC_subunits_Haris_18.04.20.csv") 
head(mrc)
gene_cl %>% 
	dplyr::filter(gene_name %in% mrc$gene_name) %>%
	dplyr::left_join(., mrc, by = "gene_name") %>%
	dplyr::group_by(complex) %>% 
	dplyr::arrange(complex, desc(PD)) %>%
	dplyr::mutate(diff = PD - HA) %>%
	dplyr::mutate(class = ifelse(class == "subunit (alias COXFA4)", "subunit", class)) %>%
	dplyr::mutate(incrPD = ifelse( PD > HA & PD > YG & diff > 0.3, TRUE, FALSE)) %>%
	dplyr::select(gene_name, HA, PD, diff, everything()) -> df
df
m <- df %>%		
	dplyr::ungroup(.) %>%
	dplyr::select(YG, HA, PD) %>%
	as.matrix(.)

rownames(m) <- df$gene_name
col_fun2  <- viridis(6)
names(col_fun2) <- df %>% pull(class) %>% unique(.)
library(wesanderson)
col_fun1  <- c("lightgrey", 
	       wes_palette(n=5, name="Darjeeling1"),
	       "darkgrey")
names(col_fun1)  <-  df %>% pull(complex) %>% unique(.)
col_fun1[2] <- "#F763E0"
ha  <-  rowAnnotation( `(PD > HA) & (PD > YG) & (diff(PD,HA) > 0.3)` = df$incrPD,
		      Complex = df$complex,
		      Class = df$class,
			col = list("Class" = col_fun2,
				   "Complex" = col_fun1,
				   `(PD > HA) & (PD > YG) & (diff(PD,HA) > 0.3)` = c("TRUE" = "black", "FALSE" = "white")))


ComplexHeatmap::Heatmap(m,
			name = "RNA ~ protein",
			row_title = "",
			right_annotation = ha,
			row_names_gp = gpar(fontsize = 8),
			row_km = 4,
			cluster_columns = F) -> p
pdf("./result_figs/oxphos.pdf", height = 14)
draw(p)
dev.off()
```

### Transcript - protein corrlation of proteasomal genes in PD

```{r}
set.seed(7)
psm <- readr::read_csv("./referenceData/proteome.csv") 
head(psm)
gene_cl %>% 
	dplyr::filter(gene_name %in% psm$gene_name) %>%
	dplyr::left_join(., psm, by = "gene_name") %>%
	dplyr::mutate(diff = abs(PD - HA)) %>%
	dplyr::mutate(subunit = ifelse(subunit == "activator subunit", subunit, paste0(subunit, " subunit"))) %>%
	dplyr::mutate(subcomplex = ifelse(is.na(subcomplex), "-", paste0("19S ",subcomplex))) %>%
	dplyr::mutate(incrPD = ifelse( PD < 0 & PD < HA & PD < YG & diff > 0.3, TRUE, FALSE)) %>%
	dplyr::select(gene_name, HA, PD, diff, everything()) -> df
df
m <- df %>%		
	dplyr::ungroup(.) %>%
	dplyr::select(YG, HA, PD) %>%
	as.matrix(.)

rownames(m) <- df$gene_name
col_fun2  <- viridis(7)
names(col_fun2) <- df %>% pull(subunit) %>% unique(.)
library(wesanderson)
col_fun1  <- wes_palette(n=5, name="Darjeeling1")
names(col_fun1)  <-  df %>% pull(category) %>% unique(.)

ha  <-  rowAnnotation( `(PD < 0) & (PD < HA) & (PD < YG) & (diff(PD,HA) > 0.3)` = df$incrPD,
		      Complex = df$category,
		      Subunit = df$subunit,
		      Subcomplex = df$subcomplex,
			col = list( Subcomplex = c("19S lid" = "lightblue", "19S base" = "darkblue", "-" = "white"),
				   `(PD < 0) & (PD < HA) & (PD < YG) & (diff(PD,HA) > 0.3)`  = c("TRUE" = "black", "FALSE" = "white")))


ComplexHeatmap::Heatmap(m,
			name = "RNA ~ protein",
			row_title = "",
			right_annotation = ha,
			row_names_gp = gpar(fontsize = 8),
			row_km = 3,
			cluster_columns = F) -> p
pdf("./result_figs/proteasome.pdf", height = 14)
draw(p)
dev.off()

df %>% 	dplyr::ungroup() %>%
	dplyr::select(PD) %>%
	dplyr::summarize(median(PD),
			 var(PD),
			 n())
```

